password_security/tests/test_signup.py

```diff
@@ -0,0 +1,160 @@
+        }
+
+        # Weak password: SignupError is raised
+        with self.assertRaises(SignupError):
+            self.env["res.users"].signup(vals)
+
+        # Stronger password: no error raised
+        vals["password"] = "asdQWE12345_3"
+        login, pwd = self.env["res.users"].signup(vals)
+
+        # check created user
+        created_user = self.env["res.users"].search([("login", "=", "test_user")])
+        self.assertEqual(login, "test_user")
+        password_write_date = created_user.password_write_date
+        self.assertTrue(password_write_date)
+
+        # Weak password: ValidationError is raised
+        with self.assertRaises(ValidationError):
+            created_user.password = "test_user_password"
+        self.assertEqual(password_write_date, created_user.password_write_date)
+
+        # Stronger password: no error raised
+        created_user.password = "!asdQWE12345_3"
+        self.assertNotEqual(password_write_date, created_user.password_write_date)
+
+    def test_04_web_auth_signup_invalid_qcontext(self):
+        """It should catch AttributeError"""
+
+        with mock.patch("odoo.http.db_filter") as db_filter:
+            db_filter.side_effect = lambda dbs, host=None: [get_db_name()]
+            with self.assertRaises(AttributeError):
+                # 'TestPasswordSecuritySignup' object has no attribute 'session'
+                self.url_open(
+                    "/web/signup",
+                    data={
+                        "csrf_token": http.Request.csrf_token(self),
+                    },
+                )
+
+    def test_05_web_auth_signup_invalid_qcontext(self):
+        """It should catch EndTestException on signup qcontext"""
+        self.session = http.root.session_store.new()
+        self.opener = Opener(self.env.cr)
+        self.opener.cookies.set("session_id", self.session.sid, domain=HOST, path="/")
+
+        with mock.patch(
+            "odoo.addons.auth_signup.controllers.main.AuthSignupHome.get_auth_signup_qcontext"
+        ) as qcontext:
+            qcontext.side_effect = EndTestException
+            with self.assertRaises(HTTPError):
+                # Catch HTTPError: 400 Client Error: BAD REQUEST
+                self.signup("jackoneill", "!asdQWE12345_3")
+
+    @mute_logger("odoo.addons.auth_signup_verify_email.controllers.main")
+    def test_06_web_auth_signup_invalid_render(self):
+        """It should render & return signup form on invalid"""
+        self.session = http.root.session_store.new()
+        self.opener = Opener(self.env.cr)
+        self.opener.cookies.set("session_id", self.session.sid, domain=HOST, path="/")
+
+        with mock.patch("odoo.http.db_filter") as db_filter:
+            db_filter.side_effect = lambda dbs, host=None: [get_db_name()]
+            # Signup: no name or partner given for new user
+            response = self.url_open(
+                "/web/signup",
+                data={
+                    "login": "test@test.com",
```

Comment:
```suggestion
                    "login": "test@test.com",
                    "password": "test123",
```
Looking at auth_signup_verify_email, it seems like it only redirects the signup for passwordless signups. So If you set a password, it should be handled by web_auth_signup and your exception handler.

Reply:
You are right, thank you!